docs/rfcs/0006-top-sites-feature.md

```diff
@@ -0,0 +1,125 @@
+---
+layout: page
+title: Introduce a Top Sites Feature
+permalink: /rfc/0006-top-sites-feature
+---
+
+* Start date: 2020-07-30
+* RFC PR: [#7931](https://github.com/mozilla-mobile/android-components/pull/7931)
+
+## Summary
+
+Create a TopSitesFeature that abstracts out the co-ordination between the multiple data storages.
+
+## Motivation
+
+A large part of the top sites implementation is spread across A-C and Fenix. There is a lot of co-ordination between the History API and the current Top Sites API to manage the final state of what sites should be displayed that can exist as a single implementation.
+
+## Guide-level explanation
+
+To begin, we should rename the `TopSitesStorage`  to `PinnedSitesStorage` and create a `TopSitesFeature` that would orchestrate the different storages and present them to the app.
+
+We can then follow the architecture of existing features by using the presenter/interactor/view model, and introduce a `TopSitesStorage` which abstract out the complexity of the multiple data sources from `PinnedSitesStorage` and `PlacesHistoryStorage`.
+
+To allow the ability for adding a top site from different parts of the app (e.g. context menus), we introduce the `TopSitesUseCases` that acts on the storage directly. If the `TopSitesFeature` is started, it will then be notified by the `Observer` to update the UI.
+
+```kotlin
+
+/**
+ * Implemented by the application for displaying onto the UI.
+ */
+interface TopSitesView {
+  /**
+   * Update the UI.
+   */
+  fun displaySites(sites: List<TopSite>)
+
+  interface Listener {
+    /**
+     * Invoked by the UI for us to add to our storage.
+     */
+    fun onAddTopSite(topSite: TopSite)
+
+    /**
+     * Invoked by the UI for us to remove from our storage.
+     */
+    fun onRemoveTopSite(topSite: TopSite)
+  }
+}
+
+/**
+ * Abstraction layer above the multiple storages.
+ */
+interface TopSitesStorage {
+
+  /**
+   * Return a unified list of top sites based on the given number of sites desired.
+   * If `includeFrecent` is true, fill in any missing top sites with frecent top site results.
+   */
+  fun getTopSites(totalNumberOfSites: Int, includeFrecent: Boolean): List<TopSite>
+
+  interface Observer {
+    /**
+     * Invoked when changes are made to the storage.
+     */
+    fun onStorageUpdated()
+  }
+}
+
+// Already exists in browser-storage-sync.
+interface PlacesHistoryStorage {
+  fun getTopFrecentSites(num: Int)
+}
+
+class DefaultTopSitesStorage(
+  val pinnedSitesStorage: PinnedSitesStorage,
+  val historyStorage: HistoryStorage? = null
+) : TopSitesStorage {
+
+  /**
+   * Merge data sources here, return a single list of top sites.
+   */
+  override fun getTopSites(numFrecent): List<TopSite>
```

Comment:
From a performance perspective, there's a problem in the current implementation on startup. The pinned sites can only be returned from an observer. As such, on startup, UI is laid out and drawn with no pinned sites, then, when the pinned sites data is ready, it's laid out again and drawn again with the pinned sites. This 1) causes an undesirable flicker and 2) layout is expensive, particularly on low-end devices so it takes longer to reach visual completeness than if this didn't exist.

I can think of two solutions:
- have a blocking fetch API that 1) we can call early on in startup to prefetch and 2) block on for the initial UI lay out and draw on **(the line I commented on already seems to address our need** but I wanted to raise our need to you just in case it changes)**
- have an async listener API that 1) we can prefetch early on in startup and 2) will return a value even when the value is empty such that we can progressively load the homescreen: we can first inflate an empty layout without the top sites, collections, etc. and, when the data is ready, lay out and draw the complex layouts all at once

Assuming the calls are fast enough, I'm not sure which is preferably from a UI perspective. If they're slow, we may prefer option 2 to get data on the screen faster.

Ideally, all the views on the homescreen uses the same style of data loading so we only have to implement one of these solutions.

CC @csadilek @MarcLeclair

Reply:
Option 2 is what is in the current proposal: we notify the `TopSitesView.displaySites` which the app can implement on an Android View or daisy chain with other listeners and collect all the data before drawing it on the screen.

This line can act like a prefetch as well (we could make it a suspend as well probably) and let the app decide when is the best time to show the initial data.